hadoop-hdds/rocks-native/src/main/patches/rocks-native.patch

```
@@ -16,592 +16,557 @@
++#include <sstream>
++#include <vector>
++
++#include "db/blob/blob_index.h"
++#include "db/memtable.h"
++#include "db/write_batch_internal.h"
++#include "options/cf_options.h"
++#include "port/port.h"
++#include "rocksdb/db.h"
++#include "rocksdb/env.h"
++#include "rocksdb/iterator.h"
++#include "rocksdb/slice_transform.h"
++#include "rocksdb/status.h"
++#include "rocksdb/table_properties.h"
++#include "rocksdb/utilities/ldb_cmd.h"
++#include "table/block_based/block.h"
++#include "table/block_based/block_based_table_builder.h"
++#include "table/block_based/block_based_table_factory.h"
++#include "table/block_based/block_builder.h"
++#include "table/format.h"
++#include "table/meta_blocks.h"
++#include "table/plain/plain_table_factory.h"
++#include "table/table_reader.h"
++#include "util/compression.h"
++#include "util/random.h"
++
++namespace ROCKSDB_NAMESPACE {
 +
-+void write(Slice &slice, FILE* file) {
-+  int size = (int)slice.size();
-+  write(slice.data(), size, file);
++RawSstFileReaderIterator::RawSstFileReaderIterator(InternalIterator* iterator,
++                             bool has_from, Slice* from_key,
++                             bool has_to, Slice* to_key)
++    : iter_(iterator),
++      ikey(new ParsedInternalKey()),
++      has_to_(has_to),
++      to_key_(to_key) {
++  if (has_from) {
++    InternalKey k;
++    k.SetMinPossibleForUserKey(*from_key);
++    iter_->Seek(k.Encode());
++  } else {
++    iter_->SeekToFirst();
++  }
++  initKey();
 +}
 +
-+void write(SequenceNumber sequenceNumber, FILE* file) {
++bool RawSstFileReaderIterator::has_next() const {
++  return iter_->Valid() && (!has_to_ ||
++                            BytewiseComparator()->Compare(
++                                getKey(), *to_key_) < 0);
++}
 +
-+  char b[8];
-+  int idx = 7;
-+  while (idx >= 0) {
-+    b[idx] = sequenceNumber % 256;
-+    sequenceNumber /= 256;
-+    idx -= 1;
++void RawSstFileReaderIterator::initKey() {
++  if (iter_->Valid()) {
```

Does `iter_->Next();` throw an error if there is no more element? If not, we should throw an error if iterator is not valid.

---

we are handling that on the java side. Let's keep it simple on the c++ side.

---

I believe `iter_->Next();` would throw an exception.